snuba/utils/streams/topics.py

```diff
@@ -1,62 +1,90 @@
+    "SNUBA_QUERIES": "snuba-queries",
+    "PROCESSED_PROFILES": "processed-profiles",
+    "PROFILES_CALL_TREE": "profiles-call-tree",
+    "INGEST_REPLAY_EVENTS": "ingest-replay-events",
+    "SNUBA_GENERIC_METRICS": "snuba-generic-metrics",
+    "SNUBA_GENERIC_METRICS_SETS_COMMIT_LOG": "snuba-generic-metrics-sets-commit-log",
+    "SNUBA_GENERIC_METRICS_DISTRIBUTIONS_COMMIT_LOG": "snuba-generic-metrics-distributions-commit-log",
+    "SNUBA_DEAD_LETTER_INSERTS": "snuba-dead-letter-inserts",
+    "SNUBA_ATTRIBUTION": "snuba-attribution",
+    "SNUBA_DEAD_LETTER_METRICS": "snuba-dead-letter-metrics",
+    "SNUBA_DEAD_LETTER_SESSIONS": "snuba-dead-letter-sessions",
+    "SNUBA_DEAD_LETTER_GENERIC_METRICS": "snuba-dead-letter-generic-metrics",
+    "SNUBA_DEAD_LETTER_REPLAYS": "snuba-dead-letter-replays",
+    "SNUBA_SESSIONS_COMMIT_LOG": "snuba-sessions-commit-log",
+    "SCHEDULED_SUBSCRIPTIONS_SESSIONS": "scheduled-subscriptions-sessions",
+    "SESSIONS_SUBSCRIPTIONS_RESULTS": "sessions-subscription-results",
+    "SNUBA_REPLAY_EVENTS": "snuba-replay-events",
+}
+
+_REGISTERED_TOPICS: dict[str, str] = {}
+
+
+class _Topic(type):
+    def __getattr__(self, attr: str) -> "Topic":
+        if attr not in _HARDCODED_TOPICS and attr not in _REGISTERED_TOPICS:
+            raise AttributeError(attr)
+        return Topic(attr.lower().replace("_", "-"))
+
+    def __iter__(self) -> Iterator[Topic]:
+        return iter(
+            Topic(value)
+            for value in {**_HARDCODED_TOPICS, **_REGISTERED_TOPICS}.values()
+        )
+
+
+class Topic(metaclass=_Topic):
+    def __init__(self, value: str):
+        self.value = value
+
+    def __hash__(self) -> int:
+        return hash(self.value)
+
+    def __eq__(self, other: Any) -> bool:
+        return isinstance(other, Topic) and other.value == self.value
+
+    def __repr__(self) -> str:
+        return f"Topic.{self.value.upper()}"
+
+
+def register_topic(key: str) -> Topic:
+    _REGISTERED_TOPICS[key.upper().replace("-", "_")] = key.lower()
+    return Topic(key)
 
 
 def get_topic_creation_config(topic: Topic) -> Mapping[str, str]:
     config = {
         Topic.EVENTS: {"message.timestamp.type": "LogAppendTime"},
         Topic.TRANSACTIONS: {"message.timestamp.type": "LogAppendTime"},
-        Topic.METRICS: {"message.timestamp.type": "LogAppendTime"},
-        Topic.PROFILES: {"message.timestamp.type": "LogAppendTime"},
-        Topic.REPLAYEVENTS: {"message.timestamp.type": "LogAppendTime"},
-        Topic.GENERIC_METRICS: {"message.timestamp.type": "LogAppendTime"},
+        Topic.SNUBA_METRICS: {"message.timestamp.type": "LogAppendTime"},
+        Topic.PROCESSED_PROFILES: {"message.timestamp.type": "LogAppendTime"},
+        Topic.INGEST_REPLAY_EVENTS: {"message.timestamp.type": "LogAppendTime"},
+        Topic.SNUBA_GENERIC_METRICS: {"message.timestamp.type": "LogAppendTime"},
```

Comment:
Dp you plan to move this somehow to config ? 
This setting is required for any storage we want to run alerts on. So that would mean we would still have to make changes to Snuba code to create a new dataset that has subscriptions enabled. 

I think you have two options. 
- move these creation settings to config.
- make this setting applied automatically when creating a topic that has subscriptions associated.

Though I think this opens another problem: should the dataset definition carry enough information to create all topics the dataset needs or should the config be split between the topic name (defined in the dataset) and the environment specific settings to create a topic (like partitions number and retention)?
I think the environment parameters should not be together with the dataset config. The dataset is the schema, it is part of the product.

@lynnagara  what do you think ?

Reply:
Can you clarify you suggestions?

1) Have the Kafka settings in the configuration.
```
topic:
  - name: snuba-metrics
  - settings:
    - message.timestamp.type: LogAppendTime
```

2) Make the setting applied automatically:
```
topic:
  - name: snuba-metrics
  - subscriptions: yes
```
and then
```
if config["subscriptions"] == "yes":
    kafka_settings["message.timestamp.type"] = "LogAppendTime"
```

Reply:
Context:
The way snuba manages Kafka topic relies a lot on them being in an enum. This was all meant and designed long before we decided to move dataset in config. So at this point they suffer of all the same limitations and constraints `StorageSetKey` have plus others. 
Like `StorageSetKey`, topics are a logical view on an environment specific resource. As such, part of them are part of the application (the logical name) and part of that needs to be overridden per environment (name and assignment to broker).
Contrarily to `StorageSetKey`, Topics are a step ahead as they carry the configuration that should be applied when the topic is created. We do not have anything that automatically configure a Clickhouse cluster when it is created. 
The topic creation configuration contains some fields that are actually environment specific (like partitions and retention) and other that the application depends on (like `LogAppendTime`. Subscriptions cannot work without). 

So it seems that treating all creation settings as environment settings would be wrong and treating all of them as application specific would be wrong.

To solve this problem without requiring anything to be hardcoded when a new dataset is added, I think we should allow the following:
- a basic sets of settings that are applied based on the fact that the storage has subscriptions. So something like your option 2. The dataset config says whether the storage has subscriptions and the topic is created with that flag on. This does not solve it for existing topics when we activate subscriptions after the fact but I believe this is not solve today either. 
- an additional set of settings to override parameters for the specific environment. This is not needed in this PR though as it is not supported today either. 

Or, in alternative, just always create topics with `LogAppendTime` @lynnagara  what do you think about always creating topics that way ?

Reply:
I think it's very important for topics to be "registered" (in some form or another) in a single place. I think Evan's option 1 is the best of the proposed ideas here and by far the closest to how it already works today.

Why I think the other options should not be pursued:
- Firstly the `subscriptions: yes` option. Topic configuration should not know about high level Snuba concepts. One day this may actually move outside Snuba as this is how different services talk to each other. The only reason we put it inside Snuba is because nothing like that exists today.
- About putting `LogAppendTime` everywhere. Well, this solves our immediate problem but closes the door to all other custom configuration. This is inflexible and seems like it would be a pain on the day we need it.

Reply:
Now, if we want to do things like Validate LogAppendTime is there when a topic is chosen for subscriptions, we can still do that inside the storage building logic. But it should not be the job of the topics definition to validate Snuba specific concepts.

Reply:
It sounds like there are two different perspectives on how to solve this problem. Whether or not we decide on a solution here, would it make sense for this to be moved to some documentation and shared for visibility? For the scope of this PR, the **function** of the Topic class is the same as the Enum. There is another [PR](https://github.com/getsentry/snuba/pull/3259) which depends on some of the `register_topic()` functionality implemented here.

Since as mentioned, logical Topics still carry configuration that is applied in code. What I'm proposing moving forward:
1. I'll write up documentation to capture this discussion and share it for visibility
2. We can more formally decide on which option to pursue
3. This will be a separate task that gets carried out during tech debt week

Does that sound sensible? Or does it make more sense for this PR to contain these changes?  @evanh @lynnagara @fpacifici

Reply:
Yes, let's do that @enochtangg . Thank you.

I think there are two underlying questions that this is touching on: How much of the `Topic` settings is logical and belongs in configuration, and how much is environment specific? Once we figure that out, where do the environment specific settings go?